pkg/globalstats/stats.go

```diff
@@ -49,45 +52,81 @@ func Incr(name string, tags ...stats.Tag) {
 	engine.Incr(name, tags...)
 }
 
+func Disable() {
+	mut.Lock()
+	defer mut.Unlock()
+
+	if flusherStop != nil {
+		close(flusherStop)
+	}
+}
+
 func Initialize(ctx context.Context, config Config) {
-	once.Do(func() {
-		if config.AppName == "" {
+	mut.Lock()
+	defer mut.Unlock()
+
+	if config.AppName == "" {
+		if len(os.Args) > 0 {
+			config.AppName = filepath.Base(os.Args[0])
+		} else {
 			config.AppName = "unknown"
 		}
-		if config.SamplePct <= 0 || config.SamplePct > 1 {
-			// by default only sample 10% of the observations
-			config.SamplePct = 0.10
+	}
+	if config.SamplePct == 0 {
+		// By default, only sample 10% of the observations.
+		config.SamplePct = 0.10
+	}
+	if config.FlushEvery == 0 {
+		config.FlushEvery = 10 * time.Second
+	}
+
+	samplePct = config.SamplePct
+
+	err := func() error {
+		if config.StatsHandler == nil {
+			return errors.New("no datadog client supplied")
+		}
+		if config.SamplePct > 1 || config.SamplePct < 0 {
+			return errors.New("sample percentage must be in the range of (0, 1]")
+		}
+		if config.FlushEvery < 0 {
+			return errors.New("flush rate must be a positive duration")
 		}
-		samplePct = config.SamplePct
-		err := func() error {
-			client := config.StatsHandler
-			if client == nil {
-				return errors.New("no datadog client supplied")
-			}
-			tags := []stats.Tag{
-				{Name: "app", Value: config.AppName},
-				{Name: "version", Value: config.CtlstoreVersion},
-			}
-			engine = stats.NewEngine(statsPrefix, client, tags...)
-			flushEvery := 10 * time.Second
-			if config.FlushEvery > 0 {
-				flushEvery = config.FlushEvery
-			}
-			go func() {
-				defer engine.Flush()
-				for {
-					select {
-					case <-time.After(flushEvery):
-						engine.Flush()
-					case <-ctx.Done():
-						return
-					}
-				}
-			}()
-			return nil
-		}()
-		if err != nil {
-			events.Log("Could not initialize ctlstore global stats: %{error}s", err)
+		if config.CtlstoreVersion == "" {
+			return errors.New("must supply the ctlstore version")
 		}
-	})
+
+		return nil
+	}()
+	if err != nil {
+		events.Log("Could not initialize ctlstore global stats: %{error}s", err)
+	}
+
+	// Stop any goroutines from any previous Initialize calls.
+	if flusherStop != nil {
+		close(flusherStop)
+	}
+	flusherStop = make(chan struct{})
+
+	tags := []stats.Tag{
+		{Name: "app", Value: config.AppName},
+		{Name: "version", Value: config.CtlstoreVersion},
+	}
+	engine = stats.NewEngine(statsPrefix, config.StatsHandler, tags...)
+
+	go flusher(ctx, config.FlushEvery)
```

Comment:
I would prefer to pass in the stop chan as an argument to `flusher`.  this seems a little racy between the goroutine that's setting the package level variable and the flow control that is dependent on it in the flusher goroutine.